modin/experimental/cloud/rpyc_proxy.py

```diff
@@ -12,34 +12,157 @@
 # governing permissions and limitations under the License.
 
 import types
+import collections
 
 import rpyc
 from rpyc.lib.compat import pickle
-from rpyc.core import netref
+from rpyc.lib import get_methods
+
+from rpyc.core import netref, AsyncResult, consts
 
 from . import get_connection
 from .meta_magic import _LOCAL_ATTRS, RemoteMeta, _KNOWN_DUALS
 
 
+def _batch_loads(items):
+    return tuple(pickle.loads(item) for item in items)
+
+
+def _tuplize(arg):
+    """turns any sequence or iterator into a flat tuple"""
+    return tuple(arg)
+
+
 class WrappingConnection(rpyc.Connection):
     def __init__(self, *a, **kw):
         super().__init__(*a, **kw)
-        self._remote_pickle_loads = None
+        self._remote_batch_loads = None
+        self._remote_cls_cache = {}
+        self._static_cache = collections.defaultdict(dict)
+        self._remote_dumps = None
+        self._remote_tuplize = None
+
+    def __wrap(self, local_obj):
+        while True:
+            # unwrap magic wrappers first
+            try:
+                local_obj = object.__getattribute__(local_obj, "__remote_end__")
+            except AttributeError:
+                break
+        # do not pickle netrefs
+        if isinstance(local_obj, netref.BaseNetref) and local_obj.____conn__ is self:
```

Comment:
Why is `if isinstance(local_obj, netref.BaseNetref):` not enough?

Reply:
added a comment